src/Container.php

```diff
@@ -0,0 +1,82 @@
+<?php
+/**
+ * Class Container
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay;
+
+use Psr\Container\ContainerInterface;
+use WCPay\Internal\DependencyManagement\DelegateContainer\WooContainer;
+use WCPay\Internal\DependencyManagement\ExtendedContainer;
+use WCPay\Internal\DependencyManagement\ServiceProvider\PaymentsServiceProvider;
+
+/**
+ * WCPay Dependency Injection Container.
+ *
+ * Wraps the ExtendedContainer implementation to only allow public access to
+ * certain methods of the internal container.
+ *
+ * The public methods of this class should be a subset of the ExtendedContainer
+ * methods, and follow their signatures. Avoid adding other methods here.
+ *
+ * During tests, the internal ExtendedContainer is available through
+ * the `wcpay_get_test_container()` method, allowing full manipulation.
+ */
+class Container implements ContainerInterface {
+	/**
+	 * Internal container instance.
+	 *
+	 * @var ExtendedContainer
+	 */
+	private $container;
+
+	/**
+	 * Initializes the container.
+	 *
+	 * Dependencies should not be provided during runtime, but will allow
+	 * mocking during tests. This is only needed for the container.
+	 *
+	 * @param WooContainer $woo_container The delegate container for WooCommerce (Optional).
+	 */
+	public function __construct( WooContainer $woo_container = null ) {
+		$this->container = new ExtendedContainer();
+
+		// Allow the container to be used as a dependency.
+		$this->container->addShared( static::class, $this );
```

Comment:
> Is this necessary if we already have `wcpay_get_container`?

There's a bit of a mental difference: The container should be used as a dependency within `src`, while `wcpay_get_container()` should be used within `includes`.

>  I would rather the container be used only once at the flow entry point and resolve all possible dependencies, in keeping with inversion of control (the individual classes should not be concerned about how they or others are created, only business logic).

TBH, I was mostly following Woo's example here, and thought it made sense.

In general, I'm 100% 👍 for IoC. However, I don't consider one class being aware of ways to instantiate another one as opposing the principle. Even though DI and IoC go hand in hand, there's a reason why dependencies are not necessarily always shared.

`WCPay\Container` class exists for a reason, and that's to limit the methods, accessible later. Did you notice that only `has` and `get` are available, but others like `add` and `extend` are not? This prevents "mid-level" classes from altering the behavior of the application, so the control remains on top.

There are plenty of cases where it makes sense for a non-shared class to have dependencies, and for those to be provided directly by the container. I've got payment states in my mind right now, wonder who suggested that pattern? 😝 

Anticipating the next question: Technically, we could have payment states as services, and pass the `$payment` object around to all methods. Yet, it also makes sense to instantiate states specifically for a given payment, and using that payment for context, together with all other needed dependencies.

In any case, consider the following: We need the container to be accessible publicly (from `includes`), and as long as `wcpay_get_container()` is accessible publicly, developers will use it even in `src` when needed.

I would rather have the container available as a dependency because that's still better to mock, similarly to a factory:

```php
class Payment {
	private $container;
	private $state;

	public function __construct( Container $container ) {
		$this->container = $container;
	}

	public function process() {
		$state = $this->container->get( SomeState::class );
		$state->set_context( $this );
		$this->state = $state; // I don't 100% like the recursion here by the way.
	}
}

class Payment_Test extends WCPAY_UnitTestCase {
	public function test_payment_transitions_to_state() {
		$mock_container = $this->createMock( Container::class );
		$mock_state     = $this->createMock( SomeState::class );
		
		$mock_container->expects( $this->once() )
			->method( 'get' )
			->with( SomeState::class )
			->willReturn( $mock_state );
		
		$sut = new Payment( $mock_container );

		$mock_state->expects( $this->once() )
			->method( 'set_context' )
			->with( $sut );

		$sut->process();
	}
}
```

---

Another note on a similar topic: Once we migrate server requests to `src`, the case would be very similar for them. Instead of using `WC_Payments::create_request()`, a class would be able to use the container as a dependency, and instantiate requests through it. Those would be mockable in the same way during tests. Of course, requests would need not to be shared.

```php
class SomeService {
	private $container;

	public function __construct( Container $container ) {
		$this->container = $container;
	}

	public function some_method() {
		$request = $this->container->get( SomeRequest::class );
		$request->set_xyz( 123 );
		$request->send();
	}
}
```

I know this is a different can of worms, but it's already open anyway.

Reply:
I have them in mind for some time this week. :)